✨ Add `bokeh` Visualization (#684)
TissueImageAnalytics/tiatoolbox: docs/visualization.rst

Visualization interface
=======================

> Launching the interface

Reviewer: Add heading numbers.

1. Installation
2. Data Format Conventions and File Structure
3. Annotation Store Examples
4. Launching the Interface
5. General UI Control Options

Reviewer 2: I wonder if we could have sphinx do this automatically. It would make edits easier and cleaner than manual numbering.

-----------------------

> Start the interface using the command::

Reviewer: Add an instruction to install the toolbox with a link to the installation page.

    tiatoolbox visualize --img-input path\to\slides --img-input path\to\overlays

alternatively just one path can be provided; in this case it is assumed that slides and overlays are in subdirectories of that provided directory called 'slides' and 'overlays' respectively::

    tiatoolbox visualize --img-input path\to\parent_of_slides_and _overlays

In the folder(s) that your command pointed to, should be the things that you want to visualize, following the conventions in the next section.

> Data format conventions/filestructure

Reviewer: ```suggestion
Data Format Conventions and File Structure
```

-------------------------------------

in the slides folder should be all the slides you want to use, and the overlays folder should contain whatever graphs, segmentations, heatmaps etc you are interesting in overlaying over the slides.

When a slide is selected in the interface, any valid overlay file that can be found that *contains the same name* (not including extension) will be available to overlay upon it.

> Segmentation:

Reviewer: ```suggestion
Segmentation
```

                    centroid: List[],
                    contour: List[List[]],
                    prob: float,
                    type: int
        ... #can add as many additional properties as we want...
                    }
    ... # other instances
    }

geojson::

    {"type":"Feature",
    "geometry":{
        "type":"Polygon",
        "coordinates":[[[21741, 49174.09],[21737.84, 49175.12],[21734.76, 49175.93],[21729.85, 49179.85],[21726.12, 49184.84],[21725.69, 49187.95],[21725.08, 49191],[21725.7, 49194.04],[21726.15, 49197.15],[21727.65, 49199.92],[21729.47, 49202.53],[21731.82, 49204.74],[21747.53, 49175.23],[21741, 49174.09]]]},
        "properties":{"object_type":"detection","isLocked":false}
    }

If your data is not in one of these formats, it is usually fairly straightforward to build an annotation store out of your model outputs.\
A small script of 6-10 lines is usually all that is required. There are example code snippets illustrating how to create an annotation store in a variety of common scenarios in the examples section.
Most use-cases should be covered in there, or something close enough that a few tweaks to a snippet will do what is needed.

> Heatmaps:

Reviewer: ```suggestion
Heatmaps
```

Graphs:
^^^^^^

Graphs can also be overlaid. These should be provided in a dictionary format, saved as a .json file.
eg::

    graph_dict = {  'edge_index': 2 x n_edges array of indices of pairs of connected nodes
		'coordinates': n x 2 array of x,y coordinates for each graph node
		}

Additional features can be added to nodes by adding extra keys to the dictionary, eg:

::

    graph_dict = {  'edge_index': 2 x n_edges array of indices of pairs of connected nodes
    'coordinates': n x 2 array of x,y coordinates for each graph node
    'feats': n x n_features array of features for each node
    'feat_names': list n_features names for each feature
    }

It will be possible to colour the nodes by these features in the interface, and the top 10 will appear in a tooltip when hovering over a node (you will have to turn on the hovertool in the small toolbar to the right of the main window to enable this, it is disabled by default.)
The display of nodes and edges can be toggled on/off independently in the right hand panel of the interface (note, edges will be turned off by default).

Annotation Store examples:
--------------------------

> patch predictions

Reviewer: ```suggestion
Patch Predictions
```

If your annotations are in a geojson format following the sort of thing QuPath would output, that should be ok. Contours stored following hovernet-style output in a .dat file should also work. An overview of the data structure in these formats is below.

Hovernet style::

    sample_dict = {nuc_id: {
                    box: List[],
                    centroid: List[],
                    contour: List[List[]],
                    prob: float,
                    type: int
        ... #can add as many additional properties as we want...
                    }
    ... # other instances
    }

geojson::

    {"type":"Feature",
    "geometry":{
        "type":"Polygon",
        "coordinates":[[[21741, 49174.09],[21737.84, 49175.12],[21734.76, 49175.93],[21729.85, 49179.85],[21726.12, 49184.84],[21725.69, 49187.95],[21725.08, 49191],[21725.7, 49194.04],[21726.15, 49197.15],[21727.65, 49199.92],[21729.47, 49202.53],[21731.82, 49204.74],[21747.53, 49175.23],[21741, 49174.09]]]},
        "properties":{"object_type":"detection","isLocked":false}
    }

> If your data is not in one of these formats, it is usually fairly straightforward to build an annotation store out of your model outputs.\

Author: there isnt a function provided for all cases. Sometimes the user will have to write a small script of your own, so this text still needs to be there

A small script of 6-10 lines is usually all that is required. There are example code snippets illustrating how to create an annotation store in a variety of common scenarios in the examples section.
Most use-cases should be covered in there, or something close enough that a few tweaks to a snippet will do what is needed.

Heatmaps
^^^^^^^^

will display a low-res heatmap in .jpg or .png format. Should be the same aspect ratio as the WSI it will be overlaid on. When creating the image, keep in mind that white regions (255,255,255) will be made transparent.

Single channel images can also be used but are not recommended; they should take values between 0 and 255 and will simply be put through a viridis colormap. 0 values will become white background.

Whole Slide Overlays
^^^^^^^^^^^^^^^^^^^^

> Can overlay multiple WSI's on top of eachother as separate layers

Reviewer: ```suggestion
Can overlay multiple WSI's on top of each other as separate layers
```

Graphs
^^^^^^

Graphs can also be overlaid. These should be provided in a dictionary format, saved as a .json file.
eg::

>     graph_dict = {  'edge_index': 2 x n_edges array of indices of pairs of connected nodes

Reviewer: Please make sure the indentations are consistent in the documentation.

>     graph_dict = {  'edge_index': 2 x n_edges array of indices of pairs of connected nodes

Reviewer: ```suggestion
    graph_dict = {  
        'edge_index': 2 x n_edges array of indices of pairs of connected nodes
```

.. _visualization:

Visualization Interface Usage
=============================

> Tiatoolbox provides a flexible visualization tool for viewing slides and overlaying associated model outputs or annotations. It is a browser-based UI built using tiatoolbox and :ref:`Bokeh <https://bokeh.org/>`. The following assumes tiatoolbox has been installed per the instructions here: :ref:`installation <installation>`.

Reviewer: ```suggestion
TIAToolbox provides a flexible visualization tool for viewing slides and overlaying associated model outputs or annotations. It is a browser-based UI built using tiatoolbox and :ref:`Bokeh <https://bokeh.org/>`. The following assumes TIAToolbox has been installed per the instructions here: :ref:`installation <installation>`.
```

    tiatoolbox visualize --img-input path\to\parent_of_slides_and_overlays

In the folder(s) that your command pointed to, should be the things that you want to visualize, following the conventions in :ref:`Data formats <data_format>`.

Launching on a remote machine
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

As the UI is browser-based, you can launch the interface on a remote machine by logging in via ssh and forwarding the relevant ports so that the visualization of the remote slides and overlays can be viewed in the browser on your local machine. For example, connect via ssh to your remote machine::

    ssh -L 5006:localhost:5006 -L 5000:localhost:5000 user@remote_machine

this will start a ssh session where the two ports the interface uses by default (5006 and 5000) are forwarded.

you can then launch the interface on the remote machine as above (tiatoolbox must be installed on the remote machine), and open the browser on your local machine. Navigate to localhost:5006 to view the interface.

.. _interface:

2. General UI Controls and Options
----------------------------------

.. image:: images/visualize_interface.png
    :width: 100%
    :align: center
    :alt: visualize interface

> The interface is split into two main sections. The left hand side contains the main window, which displays the slide and overlays (or pontentially a linked pair of slide views), and the right hand side contains a number of UI elements to control the display of the overlays.

Reviewer: ```suggestion
The interface is split into two main sections. The left hand side contains the main window, which displays the slide and overlays (or potentially a linked pair of slide views), and the right hand side contains a number of UI elements to control the display of the overlays.
```

The 'colour type by property' box allows annotations of the specified type to be coloured by a different property to the 'global' one. For example, this could be used to have all detections coloured according to their type, but for Glands, colour by some feature describing them instead (their area, for example)

Running models
^^^^^^^^^^^^^^

Regions of the image can be selected, using either a box select or points, which can be sent to a model via selecting the model in the drop-down menu and then clicking go. Available so far are hovernet and nuclick.

To save the annotations resulting from a model, or loaded from a .geojson or .dat (will be saved as a SQLiteStore .db file which will be far quicker to load) use the save button (for the moment it is just saved in a file '{slide_name}\_saved_anns.db' in the overlays folder).

Dual window mode
^^^^^^^^^^^^^^^^

.. image:: images/dual_win.png
    :width: 100%
    :align: center
    :alt: dual window example

> A second window can be opened by selecting the 'window 2' tab in the top right. This will open the currently selected slide in a second window as illustrated above. The overlay shown in each window can be controlled independently to allow comparison of different overlays, or viewing of a model output side-by-side with the unoverlaid slide, or ground truth annotations. Slide navigation will be linked between both windows.

Reviewer: ```suggestion
A second window can be opened by selecting the 'window 2' tab in the top right. This will open the currently selected slide in a second window as illustrated above. The overlay shown in each window can be controlled independently to allow comparison of different overlays, or viewing of a model output side-by-side with the raw image (slide), or ground truth annotations. Slide navigation will be linked between both windows.
```

There are settings to control how slides are loaded:

::

    {
    "initial_views": {
        "slideA": [0,19000,35000,44000],    # if a slide with specified name is opened, initial view window will be set to this
        "slideB": [44200,59100,69700,76600]
            },
    "auto_load": 1,     # if 1, upon opening a slide will also load all annotations associated with it
    "first_slide": "slideA.svs",            # initial slide to open upon launching viewer

Settings to control how annotations are displayed, including default colours for specific types, and default properties to colour by:

::

    "colour_dict": {
        "typeA": [252, 161, 3, 255],   # annotations whose 'type' property matches these, will display in the specified colour
        "typeB": [3, 252, 40, 255]
    },
    "default_cprop": "some_property",     # default property to colour annotations by
    "default_type_cprop": {               # a property to colour a specific type by
    "type": "Gland",
    "cprop": "Explanation"
    },

There are settings to control the initial values of some UI settings:

::

    "UI_settings": {

>         "blur_radius": 0,           # applies a blur to rendererd annotations

Reviewer: ```suggestion
        "blur_radius": 0,           # applies a blur to rendered annotations
```

        "edge_thickness": 0,        # thickness of boundaries drawn around annotation geometries (0=off)

>         "mapper": "jet",            # default colourmapper to use when colouring by a continuous property

Reviewer: ```suggestion
        "mapper": "jet",            # default colour mapper to use when colouring by a continuous property
```

        "max_scale": 32             # controls zoom level at which small annotations are no longer rendered (larger val->smaller
    },                              # annotations visible when zoomed out)
    "opts": {
        "edges_on": 0,              # graph edges are shown or hidden by default
        "nodes_on": 1,              # graph nodes are shown or hidden by default

>         "colourbar_on": 1,           # whether colourbar is shown below main window

Reviewer: ```suggestion
        "colourbar_on": 1,           # whether colour bar is shown below main window
```

        "hover_on": 1
    },

and the ability to toggle on or off specific UI elements:

::

    "UI_elements_1": {              # controls which UI elements are visible
        "slide_select": 1,          # slide select box
        "layer_drop": 1,            # overlay select drop down
        "slide_row": 1,             # slide alpha toggle and slider
        "overlay_row": 1,           # overlay alpha toggle and slider
        "filter_input": 1,          # filter text input box
        "cprop_input": 1,           # box to select which property to colour annotations by ('colour by' box)
        "cmap_row": 1,              # row of UI elements with colourmap select, blur, max_scale
        "type_cmap_select": 1,      # UI element to select a secondary colourmap for a specific type (i.e 'colour type by' box)
        "model_row": 0,             # UI elements to chose and run a model

>         "type_select_row": 1        # buttom group for toggling specific types of annotations on/off

Reviewer: ```suggestion
        "type_select_row": 1        # button group for toggling specific types of annotations on/off
```

.. _visualization:

Visualization Interface Usage
=============================

TIAToolbox provides a flexible visualization tool for viewing slides and overlaying associated model outputs or annotations. It is a browser-based UI built using TIAToolbox and `Bokeh <https://bokeh.org/>`_. The following assumes TIAToolbox has been installed per the instructions here: :ref:`Installation <installation>`.

1. Launching the interface
--------------------------

Start the interface using the command::

>     tiatoolbox visualize --slide-path path\to\slides --overlay-path path\to\overlays

Reviewer: ```suggestion
    tiatoolbox visualize --slide-path /path/to/slides --overlay-path /path/to/overlays
```

just to be consistent with Linux
